Overhaul Web UI

raghavsethi opened the pull request:
Here's the overhaul of the web UI I worked on over the hackathon and the last couple of weeks.

### Design Goals
- To make it easy to see how a cluster is doing
- To make it easier to find queries in the query list
- To help figure out query performance, stragglers and skew
- To bring the Web UI into 2016

### Non-goals
- Looking decent at window sizes < 1024px
- Supporting browsers that aren't Chrome, Safari or Firefox of a vintage > 3 months ago
- Premature performance optimizations

### New Features
- Cluster heads up display: Now it's easy to see how your cluster is doing, and what sort of load it's under:
  - Running, blocked and queued queries
  - Aggregate data input rates (rows and bytes)
  - Aggregate parallelism (drivers and CPU time)
  - All of these also have sparklines for easy trend analysis
- More powerful query list
  - Finer-grained stats (execution, CPU and elapsed time)
  - Powerful sorting and filtering options
  - Limit number of displayed queries to improve page performance
  - Regex search on query ID, source, user and query text
- Completely re-written query detail page
  - Query resource utilization statistics, and sparklines for parallelism, input data and memory
  - All new stage-level statistics
    - See timings, memory and task info
    - Find stragglers and outliers with bar charts of CPU time and buffers for all tasks
  - And best of all - everything on this page is live - you can see stats change as a query progresses (can also be paused)
- All UI code has been either re-written or cleaned up dramatically. Everything is now in React.

### Regressions
- The new table framework doesn't support sticky headers out of the box (will add it back eventually)
- Sorting the tasks table by multiple columns sequentially does not maintain previous sort orders
- Removed the stage view in the query detail page (it seemed to be universally despised)
- The density of information on the cluster overview page seems like a potential regression to me, but the demand for more information won the battle.
- There are about 5x more bells and whistles than before - which means that I'm going to be fixing edge cases for weeks.

### Feedback Required
cc @martint @dain @haozhun @electrum @nileema 
- I've made dozens of copy and design decisions, please let me know if (a) you feel strongly about something or (b) it's easy to change.
- I need to figure out how to roll this out in a way that we're not stuck with a broken UI in between releases. My current idea is to have a banner at the top of the old UI with a link to the new one first. Then I would change it to the new UI and put a link to the old one. Then I would remove the old one completely. Is this too complex? Can I just bite the bullet and replace the old UI?

As an aside, I've tried to avoid adding dependencies, especially into the build process. This means that I rejected all libraries that really wanted me to have Webpack or Browserify, or ones that the only way to get a piece of JS out of was to install a CoffeeScript-JSX Compiler (!!). This also means that we're using JSX, which is not great for performance, but I'm going to punt on compiling stuff until we know there's a perf problem.

electrum commented:
Can we keep the color logo?

electrum commented:
Please split this up to add the new vendor resources as a separate commit. This will make it easier to review and look at later.

raghavsethi commented:
Split into two commits that will be squashed before merge.

- Changed to color logo
- Added histograms
- Made query page sparklines show instantaneous rather than average stats

cberner commented:
Can you split up the first commit into 1) make copy of UI under /old/, 2) add new libraries? That will make it easier to review, as right now there's just a 81k line commit and it's not clear what's happening :) Also the commits need better titles and descriptions

raghavsethi commented:
Restructured the PR to make it easier to review, also made a whole bunch of perf improvements.

cberner commented on presto-main/src/main/resources/webapp/vendor/bootstrap/config.json:
```diff
@@ -0,0 +1,435 @@
+{
```
List in the commit message what libraries are being added.

cberner replied:
Also, are you sure all the necessary license files are included? It seems like the license for Solarized, at least, is missing: https://github.com/altercation/solarized/blob/master/LICENSE

raghavsethi commented:
Listed libraries in the commit, and added license files for all libraries that were missing them.

cberner commented on presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java:
```diff
@@ -192,6 +204,36 @@ public Duration getElapsedTime()
+    }
+
+    @JsonProperty
+    public long getExecutionTimeMillis()
```
Why do we have both `getExecutionTime` and `getExecutionTimeMillis`?

raghavsethi replied:
To maintain compatibility for the old UI. Should I change the old one instead?

cberner replied:
I'm a bit confused. Why do we need both for backwards compatibility? Both of these methods are new

raghavsethi replied:
Sorry I confused this with another method in a an earlier version of the PR. I'll get rid of the millis code.

ArturGajowy commented:
I propose using http://www.webjars.org/ for managing frontend dependencies

ArturGajowy commented on presto-main/src/main/resources/webapp/assets/cluster.js:
```diff
@@ -0,0 +1,740 @@
+        document.getElementById('cluster-hud')
+);
+
+var Query = React.createClass({
```
Maybe let's split this file so that each react component has its own file?

ArturGajowy replied:
This also concerns assets/query.js

cberner replied:
Ya, I agree. This file is pretty long

raghavsethi replied:
Breaking this up into multiple files that have dependencies on each other means that we will have to either do some sort of ugly hack to pass globals around or use something like Webpack. I think having files be longer than we like is the lesser of the two evils.

cberner commented on presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:
```diff
@@ -0,0 +1,179 @@
+                    byteInputRate += query.getQueryStats().getProcessedInputDataSize().toBytes() / totalExecutionTimeSeconds;
+                    rowInputRate += query.getQueryStats().getProcessedInputPositions() / totalExecutionTimeSeconds;
+                    cpuTimeRate += (query.getQueryStats().getTotalCpuTime().toMillis() / 1000.0) / totalExecutionTimeSeconds;
+                    memoryReservation += query.getQueryStats().getTotalMemoryReservation().toBytes();
```
Instead of total "user" memory. It might be more useful to display free distributed memory in the general pool & system pool, along with min local memory on a node for general & system

sopel39 replied:
Or maybe also a % of free memory

raghavsethi replied:
This seems like it needs Java code changes and more thought about how to get that information. I'm going to create a separate issue to track that feature request.

cberner replied:
Should all 5 of these be inside the `if` block? It seems like only the first 3 should be

cberner replied:
I think you missed this comment

cberner commented on presto-main/src/main/java/com/facebook/presto/server/QueryResource.java:
```diff
@@ -68,16 +69,16 @@ public QueryResource(QueryManager queryManager, AccessControl accessControl, Ses
 
     @GET
-    public List<BasicQueryInfo> getAllQueryInfo()
+    public Map<String, BasicQueryInfo> getAllQueryInfo()
```
Do we need this change? It seems like the client can just index the infos on the client side, if they want a map

cberner commented on presto-main/src/main/resources/webapp/assets/query.js:
```diff
@@ -0,0 +1,1087 @@
+    render: function() {
+        var tasks = this.getTasks(this.props.outputStage);
+        var renderedTasks = tasks.map(function (task) {
+            var t = task;
```
avoid single letter variables

cberner commented on presto-main/src/main/resources/webapp/assets/query.js:
```diff
@@ -0,0 +1,1087 @@
+                elapsedTime = Date.now() - Date.parse(t.stats.createTime);
+            }
+            return (
+                    <Tr key={t.taskStatus.taskId}>
```
Seems like all the other tags are lower case, so I would stick with that for "Tr" and "Td" too

cberner commented on presto-main/src/main/resources/webapp/assets/query.js:
```diff
@@ -0,0 +1,1087 @@
+        }.bind(this));
+
+        return (
+            <Table id="tasks" className="table table-striped sortable" sortable={true} defaultSort={{column: 'id', direction: 'asc'}}>
```
lower case

cberner commented on presto-main/src/main/resources/webapp/assets/query.js:
```diff
@@ -0,0 +1,1087 @@
+
+        return (
+            <Table id="tasks" className="table table-striped sortable" sortable={true} defaultSort={{column: 'id', direction: 'asc'}}>
+                <Thead>
```
lower case for all of these

raghavsethi replied:
This casing comes from the Reactable library

cberner commented on presto-main/src/main/resources/webapp/assets/query.js:
```diff
@@ -0,0 +1,1087 @@
+        $(histogramId).sparkline(histogramData, stageHistogramProperties);
+    },
+    componentDidUpdate: function() {
+        var s = this.props.stage;
```
avoid single letter variable names

cberner commented on presto-main/src/main/resources/webapp/assets/query.js:
```diff
@@ -0,0 +1,1087 @@
+            return previousValue + currentValue;
+        });
+
+        return (
```
This block of HTML is pretty long. Maybe there's a way to split it up a bit more?

raghavsethi replied:
I believe that splitting this up makes it harder to maintain. All the code is essentially just layout, and putting stuff in separate methods would make it harder to reason about the resultant DOM.

cberner commented on presto-main/src/main/resources/webapp/assets/utils.js:
```diff
@@ -0,0 +1,374 @@
+    return valuesArray.concat([value]).slice(Math.max(valuesArray.length - MAX_HISTORY, 0));
+}
+
+function addEwmaToHistory (value, valuesArray) {
```
What's a Ewma? Is there a better name for this?

cberner commented:
Seems like there's something wrong with the time rendering:
![screen shot 2016-05-24 at 11 04 16 am](https://cloud.githubusercontent.com/assets/842688/15514764/847f709c-219f-11e6-826f-87e802ec7562.png)

cberner commented:
I found it hard to tell which buttons were selected:
![screen shot 2016-05-24 at 11 06 42 am](https://cloud.githubusercontent.com/assets/842688/15514793/a9390eac-219f-11e6-9fc3-374800ba81dd.png)

Maybe check with a few others to see if it's just me?

cberner commented:
Cumulative memory looks wrong:
![screen shot 2016-05-24 at 11 09 26 am](https://cloud.githubusercontent.com/assets/842688/15514862/071e8196-21a0-11e6-8c1f-e0b3ea55628e.png)

How can it be non-zero when the peak is zero?

cberner commented:
I think displaying the current memory usage for each query on the main query list would be useful, and being able to sort by that would be great. A somewhat common case is wanting to find the biggest query by memory so that you can kill it manually to free up the cluster

cberner commented:
Back to you for some comments

sopel39 commented on presto-main/src/main/resources/webapp/assets/cluster.js:
```diff
@@ -0,0 +1,740 @@
+
+        $('[data-toggle="tooltip"]').tooltip();
+    },
+    render: function() {
```
Is it possible to separate renedered templates from classes in React? If so, please move templates to separate files. It would increase code readability.

raghavsethi replied:
React doesn't really do templates. AFAIK this is idiomatic React.

sopel39 commented on presto-main/src/main/resources/webapp/assets/cluster.js:
```diff
@@ -0,0 +1,740 @@
+
+        var completedDrivers = query.completedDrivers;
+        var runningDrivers = query.runningDrivers;
+        var queuedDrivers = query.queuedDrivers;
```
I would guess it should be possible to precompute all variables required for rendering and pass them to template engine. Is it possible in React?

raghavsethi replied:
I've either precomputed everything that is complex or factored it out into separate functions.

sopel39 commented on presto-main/src/main/resources/webapp/old/units.js:
```diff
@@ -16,8 +16,8 @@ function formatDuration(millis) {
 
-function formatDataSize(bytes) {
-    var unit = "B";
+function formatDataSize(bytes, defaultUnit = "B") {
```
How is this used? What if `defaultUnit != "B"` and `bytes < 1024`. What is example output?

raghavsethi replied:
This change shouldn't have been committed at all. Thanks for pointing it out :)

raghavsethi commented:
@cberner back to you. It's also running in the test cluster so you can play with it.

sopel39 commented on presto-main/src/main/resources/webapp/assets/utils.js:
```diff
@@ -0,0 +1,387 @@
+
+// display at most 5 minutes worth of data on the sparklines
+var MAX_HISTORY = 60 * 5;
+ // alpha param of exponentially weighted modving average. picked arbitrarily - lower values means more smoothness
```
modving -> moving

sopel39 commented on presto-main/src/main/resources/webapp/assets/utils.js:
```diff
@@ -0,0 +1,387 @@
+
+function taskIdToSortableString(id) {
+    var taskIdArr = removeQueryId(id).split(".");
+    return taskIdArr.map( function(numStr) { return new Array(5 - numStr.length).join("0") + numStr }).join("|");
```
why 5? This formula seems a bit complicated.
Please add an example `id` so reader can understand this transformation

cberner replied:
Can React take a custom comparator instead? I think that would make it easier to understand

sopel39 commented on presto-main/src/main/resources/webapp/assets/utils.js:
```diff
@@ -0,0 +1,387 @@
+    return precisionRound(count) + unit;
+}
+
+function formatDataSizeBytes(size) {
```
Is it necessary to have almost duplicate function of `formatDataSize`?

sopel39 commented:
Possible UI bugs when running `explain analyze select * from tpch.sf1.orders,tpch.sf10.orders;` query. Browser is Chrome

When clicking `Live Plan` on query page:
redirected to: `http://localhost:8080/ui/plan?20160602_111303_00011_cn56s`
error:
```
http://localhost:8080/vendor/dagre-d3/dagre-d3-0.4.14.js Failed to load resource: the server responded with a status of 404 (Not Found)
http://localhost:8080/units.js Failed to load resource: the server responded with a status of 404 (Not Found)
plan?20160602_111303_00011_cn56s:212 Uncaught ReferenceError: dagreD3 is not defined(anonymous function) @ plan?20160602_111303_00011_cn56s:212
http://localhost:8080/favicon.ico Failed to load resource: the server responded with a status of 404 (Not Found)
```

When clicking `Split Timeline`
redirected to: `http://localhost:8080/timeline.html?20160602_111303_00011_cn56s`
error:
```
timeline.html?20160602_111303_00011_cn56s:124 Uncaught TypeError: Cannot read property 'substring' of undefined
```
I think ` task.taskId` is `undefined`

sopel39 commented:
![zrzut ekranu 2016-06-02 o 13 31 48](https://cloud.githubusercontent.com/assets/5989165/15743512/557d9fb6-28c6-11e6-96e2-c382fd751105.png)

I noticed that there is no UI way of going back from deeper page (e.g. query page) to more top level page (e.g. main page). The only way to go to top level page is by using browser `back` button.

I would suggest to add breadcrumb navigation bar at the top of each page. It would basically show clickable link to the higher level pages, e.g:
`Queries -> Query 123 -> Slice Timeline`.
See http://marketingland.com/breadcrumb-links-good-user-experience-yes-97848

cberner commented on presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java:
```diff
@@ -70,8 +76,13 @@ public BasicQueryInfo(
             @JsonProperty("endTime") DateTime endTime,
             @JsonProperty("createTime") DateTime createTime,
+            @JsonProperty("currentMemory") DataSize currentMemory,
+            @JsonProperty("currentMemory") DataSize peakMemory,
```
typo

cberner commented on presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java:
```diff
@@ -186,9 +207,21 @@ public String getQuery()
+    }
+
+    @JsonProperty
+    public long getElapsedTimeMillis()
```
This doesn't look like it's going to round trip anymore. On the other hand, I'm not sure we ever deserialize these. Either way though, either remove the JSON constructor or fix these so that it round-trips

cberner commented on presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:
```diff
@@ -0,0 +1,179 @@
+                }
+            }
+
+            if (!query.getState().equals(QueryState.FAILED) && !query.getState().equals(QueryState.FINISHED)) {
```
I think `!query.getState().isDone()` would be better

cberner commented on presto-main/src/main/resources/webapp/assets/presto.css:
```diff
@@ -0,0 +1,551 @@
+}
+
+/**
+The MIT License (MIT)
```
Why is this here? If some of the code is copied from somewhere else, I would move it to a separate file

cberner commented on presto-main/src/main/resources/webapp/query.html:
```diff
@@ -0,0 +1,106 @@
+            <div id="navbar" class="navbar-collapse collapse">
+                <ul class="nav navbar-nav navbar-right">
+                    <li>
+                            <span class="navbar-cluster-info">
```
Spacing looks off here

cberner commented on presto-main/src/main/resources/webapp/query.html:
```diff
@@ -0,0 +1,106 @@
+                            </span>
+                    </li>
+                    <li>
+                            <span class="navbar-cluster-info">
```
and here

cberner commented on presto-main/src/main/resources/webapp/index.html:
```diff
@@ -0,0 +1,111 @@
+            <div id="navbar" class="navbar-collapse collapse">
+                <ul class="nav navbar-nav navbar-right">
+                    <li>
+                            <span class="navbar-cluster-info">
```
Spacing

cberner commented on presto-main/src/main/resources/webapp/index.html:
```diff
@@ -0,0 +1,111 @@
+                            </span>
+                    </li>
+                    <li>
+                            <span class="navbar-cluster-info">
```
and here

cberner commented:
Back to you for comments

raghavsethi commented:
@sopel39 thanks for your comments. I'm going to leave breadcrumb navigation as something we can explore later, you can click the Presto logo to go back to the cluster overview page for now. #5129 is already tracking the split timeline bug.

@cberner back to you.

cberner commented on presto-main/src/main/resources/com/facebook/presto/server/plan.html:
```diff
@@ -4,7 +4,7 @@
     <script src="/vendor/d3/d3-3.3.4.js"></script>
     <script src="/vendor/dagre-d3/dagre-d3-0.4.14.js"></script>
-    <script src="/units.js"></script>
+    <script src="/assets/utils.js"></script>
```
Seems like this change should be in a different commit

cberner commented on presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:
```diff
@@ -0,0 +1,179 @@
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/")
```
Make this `/v1/cluster` in the style of `QueryResource`

cberner commented on presto-main/src/main/resources/webapp/assets/cluster-hud.js:
```diff
@@ -0,0 +1,259 @@
+    },
+    componentDidUpdate: function() {
+        // prevent multiple calls to setState within the refresh interval from re-rendering sparklines/charts
+        if (this.state.lastRender == null || (Date.now() - this.state.lastRender) >= 1000) {
```
This `lastRender` looks hacky. Why do we need it?

cberner commented on presto-main/src/main/resources/webapp/assets/loader.css:
```diff
@@ -0,0 +1,64 @@
+/**
+The MIT License (MIT)
+
+Copyright (c) 2014 Luke Haas
```
Should this file be in the vendor directory? I assume it came from somewhere else

cberner commented on presto-main/src/main/resources/webapp/assets/query-list.js:
```diff
@@ -0,0 +1,509 @@
+            case "FINISHED":
+                progress = 100;
+                runningDrivers = 0;
+                queuedDrivers = 0;
```
Add `break;`

cberner commented on presto-main/src/main/resources/webapp/assets/query-list.js:
```diff
@@ -0,0 +1,509 @@
+                </span>
+                <span className="tinystat" data-toggle="tooltip" data-placement="top" title="Queued splits">
+                    <span className="glyphicon glyphicon-pause" style={GLYPHICON_HIGHLIGHT}></span>&nbsp;&nbsp;
+                    { queuedDrivers }
```
You're inconsistent about having or not having spaces before/after the braces. Make them consistent

cberner commented on presto-main/src/main/resources/webapp/assets/utils.js:
```diff
@@ -0,0 +1,367 @@
+        return valuesArray.concat([value]);;
+    }
+
+    var ewma = (value * EWMA_ALPHA) + (valuesArray[valuesArray.length - 1] * (1 - EWMA_ALPHA));
```
avoid abbreviations. Maybe `movingAverage`?

cberner commented on presto-main/src/main/resources/webapp/assets/utils.js:
```diff
@@ -0,0 +1,367 @@
+}
+
+function getHostname(hostname) {
+    var s = new URL(hostname).hostname;
```
avoid single letter variables

cberner commented:
Had a few more comments

raghavsethi commented:
Back to you

cberner commented:
Have @dain, @martint, or @electrum look at the REST API changes. The rest looks good

electrum commented on presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java:
```diff
@@ -239,6 +241,16 @@ public Duration getQueuedTime()
+    {
+        if (queuedTime == null) {
+            // counter-intuitively, this means that the query is still queued
+            return new Duration(0, TimeUnit.NANOSECONDS);
```
Static import `NANOSECONDS`

electrum commented on presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java:
```diff
@@ -227,6 +257,24 @@ public DateTime getCreateTime()
+    }
+
+    @JsonProperty
+    public DataSize getCurrentMemory()
```
Make these bytes

electrum commented on presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:
```diff
@@ -0,0 +1,178 @@
+    @Inject
+    public ClusterStatsResource(InternalNodeManager nodeManager, QueryManager queryManager)
+    {
+        this.nodeManager = nodeManager;
```
Check for null

electrum commented on presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:
```diff
@@ -0,0 +1,178 @@
+@Path("/v1/cluster")
+public class ClusterStatsResource
+{
+    private final InternalNodeManager nodeManager;
```
These can all be `NodeManager`

electrum commented on presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:
```diff
@@ -0,0 +1,178 @@
+        double cpuTimeRate = 0;
+
+        for (QueryInfo query : queryManager.getAllQueryInfo()) {
+            if (query.getState().equals(QueryState.QUEUED)) {
```
Enums can use `==`

electrum commented on presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:
```diff
@@ -0,0 +1,178 @@
+            if (query.getState().equals(QueryState.QUEUED)) {
+                queuedQueries++;
+            }
+            if (query.getState().equals(QueryState.RUNNING)) {
```
This can be `else`

electrum commented on presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:
```diff
@@ -0,0 +1,178 @@
+            }
+
+            if (!query.getState().isDone()) {
+                double totalExecutionTimeSeconds = query.getQueryStats().getElapsedTime().toMillis() / 1000.0;
```
I think this should be `.getElapsedTime().getValue(SECONDS)`

electrum commented on presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:
```diff
@@ -0,0 +1,178 @@
+                if (totalExecutionTimeSeconds != 0) {
+                    byteInputRate += query.getQueryStats().getProcessedInputDataSize().toBytes() / totalExecutionTimeSeconds;
+                    rowInputRate += query.getQueryStats().getProcessedInputPositions() / totalExecutionTimeSeconds;
+                    cpuTimeRate += (query.getQueryStats().getTotalCpuTime().toMillis() / 1000.0) / totalExecutionTimeSeconds;
```
Same

electrum commented on presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:
```diff
@@ -0,0 +1,178 @@
+import com.facebook.presto.spi.NodeState;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.inject.Inject;
```
Use `javax.inject` instead
